src/rlookup.c

```diff
@@ -86,8 +53,195 @@ const FieldSpec *findFieldInSpecCache(const RLookup *lookup, const char *name) {
+  if (flags & RLOOKUP_F_NAMEALLOC && !(key->flags & RLOOKUP_F_NAMEALLOC)) {
+    // case 1
+    key->name = rm_strndup(key->name, key->name_len);
+  } else if (!(flags & RLOOKUP_F_NAMEALLOC) && (key->flags & RLOOKUP_F_NAMEALLOC)) {
+    // case 4
+    flags |= RLOOKUP_F_NAMEALLOC;
+  }
+
+  key->path = key->name;
+  key->flags = flags & ~RLOOKUP_TRANSIENT_FLAGS;
+}
+
+static void setKeyByFieldSpec(RLookupKey *key, const FieldSpec *fs) {
+  key->flags |= RLOOKUP_F_DOCSRC | RLOOKUP_F_SCHEMASRC;
+  if (FieldSpec_IsSortable(fs)) {
+    key->flags |= RLOOKUP_F_SVSRC;
+    key->svidx = fs->sortIdx;
+
+    if (FieldSpec_IsUnf(fs)) {
+      // If the field is sortable and not normalized (UNF), the available data in the
+      // sorting vector is the same as the data in the document.
+      key->flags |= RLOOKUP_F_ISLOADED;
+    }
+  }
+  if (FIELD_IS(fs, INDEXFLD_T_NUMERIC)) {
+    key->flags |= RLOOKUP_T_NUMERIC;
+  }
+}
+
+// Gets a key from the schema if the field is sortable (so its data is available).
+static RLookupKey *genKeyFromSpec(RLookup *lookup, const char *name, size_t name_len, int flags) {
+  const FieldSpec *fs = findFieldInSpecCache(lookup, name);
+  if(!fs || !FieldSpec_IsSortable(fs)) {
+    return NULL;
+  }
+
+  RLookupKey *key = createNewKey(lookup, name, name_len);
+  setKeyByFlags(key, flags);
+
+  setKeyByFieldSpec(key, fs);
+  return key;
+}
+
+static void setLoadedKey(RLookup *lookup, RLookupKey *key, const char *name, size_t name_len) {
+  const FieldSpec *fs = findFieldInSpecCache(lookup, name);
+
+  // Whether the key is in the schema or not, it is considered loaded and doc-source.
+  key->flags |= RLOOKUP_F_ISLOADED | RLOOKUP_F_DOCSRC;
+
+  if (fs) {
+    setKeyByFieldSpec(key, fs);
+  }
+}
+
+static RLookupKey *RLookup_FindKey(RLookup *lookup, const char *name, size_t n) {
+  for (RLookupKey *kk = lookup->head; kk; kk = kk->next) {
+    // match `name` to the name/path of the field
+    if (kk->name_len == n && !strncmp(kk->name, name, kk->name_len)) {
+      return kk;
+    }
+  }
+  return NULL;
+}
+
+RLookupKey *RLookup_GetKeyEx(RLookup *lookup, const char *name, size_t n, RLookupMode mode, int flags) {
+  // remove all flags that are not relevant to getting a key
+  flags &= RLOOKUP_GET_KEY_FLAGS;
+  // First, look for the key in the lookup table for an existing key with the same name
+  RLookupKey *key = RLookup_FindKey(lookup, name, n);
+
+  switch (mode) {
+  // 1. if the key is already loaded, or it has created by earlier RP for writing, return NULL
+  // 2. try generating a key from the schema, including all relevant meta data
+  // 3. if the key is not found in the schema, it might be in the document.
+  //    create a new key with the name of the field, and mark it as doc-source loaded.
+  case RLOOKUP_M_LOAD:
+    if (lookup->spcache == NULL) {
+      // If there is no schema, we can't load the key - fail (error?)
+      return NULL;
```

Comment:
why is this checked here?
if having no specache is an error it's not this function's responsibility to raise the flag

Reply:
An RLookup with no schema is an RLookup that cannot load anything. This is tested today by verifying that the current RL is the first RL. This is an additional verification, and I think it is a good one as we now distinguish between getting a key for loading and for other stuff.